cpp/src/io/parquet/column_reader.cu

```diff
@@ -0,0 +1,725 @@
+                  thrust::raw_pointer_cast(work_space_vector.data());
+                thrust::device_vector<T> d_values_in(values,
+                                                     values + total_values);
+                compact_to_sparse_for_nulls(
+                  thrust::raw_pointer_cast(d_values_in.data()),
+                  values,
+                  definition_levels,
+                  max_definition_level,
+                  total_values,
+                  work_space);
+            }
+        }
+        *levels_read = num_def_levels;
+        *nulls_count = null_count;
+    } else {
+        total_values = _ReadValues(current_decoder_, batch_size, values);
+        _TurnBitOnForValids(total_values, valid_bits, valid_bits_offset);
+        *nulls_count = 0;
+        *levels_read = total_values;
+    }
+
+    ConsumeBufferedValues(*levels_read);
+
+    return total_values;
+}
+
+template <class DataType>
+inline std::int64_t
+ColumnReader<DataType>::ReadBatch(std::int64_t  batch_size,
+                                  std::int16_t *def_levels,
+                                  std::int16_t *rep_levels,
+                                  T *           values,
+                                  std::int64_t *values_read) {
+    // assert(rep_levels == nullptr);
+    if (!HasNext()) {
+        *values_read = 0;
+        return 0;
+    }
+    batch_size =
+      std::min(batch_size, num_buffered_values_ - num_decoded_values_);
+
+    std::int64_t num_def_levels = 0;
+
+    std::int64_t values_to_read = 0;
+
+    if (descr_->max_definition_level() > 0 && def_levels) {
+        num_def_levels = ReadDefinitionLevels(batch_size, def_levels);
+        int result     = thrust::count(thrust::device_pointer_cast(def_levels),
+                                   thrust::device_pointer_cast(def_levels)
+                                     + num_def_levels,
+                                   descr_->max_definition_level());
+        values_to_read = result;
+    } else {
+        values_to_read = batch_size;
+    }
+
+    *values_read = _ReadValues(current_decoder_, values_to_read, values);
+    std::int64_t total_values = std::max(num_def_levels, *values_read);
+    ConsumeBufferedValues(total_values);
+
+    return total_values;
+}
+
+template <class DataType>
+struct ParquetTraits {};
+
+#define TYPE_TRAITS_FACTORY(ParquetType, GdfDType)                             \
+    template <>                                                                \
+    struct ParquetTraits<ParquetType> {                                        \
+        static constexpr gdf_dtype gdfDType = GdfDType;                        \
+    }
+
+TYPE_TRAITS_FACTORY(::parquet::BooleanType, GDF_INT8);
+TYPE_TRAITS_FACTORY(::parquet::Int32Type, GDF_INT32);
+TYPE_TRAITS_FACTORY(::parquet::Int64Type, GDF_INT64);
+TYPE_TRAITS_FACTORY(::parquet::FloatType, GDF_FLOAT32);
+TYPE_TRAITS_FACTORY(::parquet::DoubleType, GDF_FLOAT64);
```

Comment:
Does this PR include BYTE_ARRAY support? I don't see that type anywhere in the diff.